apps/remote_control/lib/lexical/remote_control/commands/rename.ex

```diff
@@ -0,0 +1,131 @@
+  # `DidChange`, `DidSave`, etc., which will trigger expensive actions like compiling the entire project.
+  # Therefore, we need this module to tell us if lexical is currently in the process of renaming.
+
+  alias Lexical.RemoteControl.Api.Messages
+  require Logger
+  import Messages
+
+  defmodule State do
+    defstruct uri_with_expected_operation: %{}, on_update_progess: nil, on_complete: nil
+
+    def new(uri_with_expected_operation, progress_functions) do
+      {on_update_progess, on_complete} = progress_functions
+
+      %__MODULE__{
+        uri_with_expected_operation: uri_with_expected_operation,
+        on_update_progess: on_update_progess,
+        on_complete: on_complete
+      }
+    end
+
+    def update_progress(%__MODULE__{} = state, file_changed(uri: uri)) do
+      update_progress(state, uri, file_changed(uri: uri))
+    end
+
+    def update_progress(%__MODULE__{} = state, file_saved(uri: uri)) do
+      update_progress(state, uri, file_saved(uri: uri))
+    end
+
+    defp update_progress(%__MODULE__{} = state, uri, message) do
+      new_uri_with_expected_operation =
+        maybe_pop_expected_operation(
+          state.uri_with_expected_operation,
+          uri,
+          message,
+          state.on_update_progess
+        )
+
+      if new_uri_with_expected_operation == %{} do
+        state.on_complete.()
+      end
+
+      %__MODULE__{state | uri_with_expected_operation: new_uri_with_expected_operation}
+    end
+
+    def in_progress?(%__MODULE__{} = state) do
+      state.uri_with_expected_operation != %{}
+    end
+
+    def maybe_pop_expected_operation(uri_to_operation, uri, message, on_update_progess) do
+      case uri_to_operation do
+        %{^uri => ^message} ->
+          on_update_progess.(1, "")
+          Map.delete(uri_to_operation, uri)
+
+        _ ->
+          uri_to_operation
+      end
+    end
+  end
+
+  alias Lexical.RemoteControl.Api.Proxy
+  use GenServer
+
+  def child_spec(uri_with_expected_operation, progress_functions) do
+    %{
+      id: __MODULE__,
+      start: {__MODULE__, :start_link, [uri_with_expected_operation, progress_functions]},
+      restart: :transient
+    }
+  end
+
+  def start_link(uri_with_expected_operation, progress_functions) do
+    state = State.new(uri_with_expected_operation, progress_functions)
+    GenServer.start_link(__MODULE__, state, name: __MODULE__)
+  end
+
+  @impl true
+  def init(state) do
+    {:ok, state, {:continue, :start_buffering}}
+  end
+
+  @spec update_progress(Messages.file_changed() | Messages.file_saved()) :: :ok
+  # NOTE:
+  # This GenServer cannot simply subscribe to messages and then update the rename status.
+  # Instead, it should call this function to synchronously update the status,
+  # thus preventing failures due to latency issues.
+  def update_progress(message) do
+    if in_progress?() do
+      GenServer.cast(__MODULE__, {:update_progress, message})
+    else
+      :ok
+    end
+  end
+
+  def in_progress? do
```

Comment:
i think this function isn't necessary. Either a rename is in progress, or it's not. If it is, you can return an error, and if it's not, a new rename can be started. It's safer to do as much as you can inside the genserver.

Reply:
(in other words, this function is extremely susceptible to races)